File: connectors-editors/msp-editor/src/main/java/com/taskadapter/connector/msp/editor/MSPEditorFactory.java
package com.taskadapter.connector.msp.editor;

import com.taskadapter.common.ui.FieldMapping;
import com.taskadapter.connector.common.FileNameGenerator;
import com.taskadapter.connector.definition.ConnectorSetup;
import com.taskadapter.connector.definition.FileSetup;
import com.taskadapter.connector.definition.exceptions.BadConfigException;
import com.taskadapter.connector.msp.MSPConfig;
import com.taskadapter.connector.msp.MSPConnector;
import com.taskadapter.connector.msp.MSPFileReader;
import com.taskadapter.connector.msp.MSPUtils;
import com.taskadapter.connector.msp.UnsupportedRelationType;
import com.taskadapter.connector.msp.editor.error.InputFileNameNotSetException;
import com.taskadapter.web.ConnectorSetupPanel;
import com.taskadapter.web.DroppingNotSupportedException;
import com.taskadapter.web.PluginEditorFactory;
import com.taskadapter.web.configeditor.file.FileProcessingResult;
import com.taskadapter.web.configeditor.file.LocalModeFilePanel;
import com.taskadapter.web.configeditor.file.ServerModeFilePanel;
import com.taskadapter.web.data.Messages;
import com.taskadapter.web.service.Sandbox;
import com.taskadapter.web.uiapi.DefaultSavableComponent;
import com.taskadapter.web.uiapi.SavableComponent;
import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;

import java.io.File;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

public class MSPEditorFactory implements PluginEditorFactory<MSPConfig, FileSetup> {
    private static final String BUNDLE_NAME = "com.taskadapter.connector.msp.messages";
    private static final String LABEL_DESCRIPTION_TEXT = "Description:";
    private static final String LABEL_TOOLTIP = "Text to show for this connector on 'Export' button. Enter any text.";
    private static final String UPLOAD_SUCCESS = "Upload success";
    private static final String UPLOAD_MPP_SUCCESS = "File uploaded and successfully converted to XML";
    private static final String SAVE_FILE_FAILED = "Save file error"; // error of saving after upload

    private static final Messages messages = new Messages(BUNDLE_NAME);

    @Override
    public String formatError(Throwable e) {
        if (e instanceof UnsupportedRelationType) {
            return messages.format("errors.unsupportedRelation",
                    messages.get("relations." + ((UnsupportedRelationType) e).getRelationType()));
        }
        if (e instanceof InputFileNameNotSetException) {
            return messages.get("error.inputFileNameNotSet");
        }
        return e.getMessage();
    }

    @Override
    public SavableComponent getMiniPanelContents(Sandbox sandbox, MSPConfig config, FileSetup setup) {
        var layout = new VerticalLayout();
        layout.setMargin(true);

        var infoReadOnlyPanel = new MSPInfoPanel();
        infoReadOnlyPanel.setHeight("152px");

        layout.add(infoReadOnlyPanel);
        // TODO 14 update the save function
        return new DefaultSavableComponent(layout, () -> {
            return true;
        });
    }

    @Override
    public boolean isWebConnector() {
        return false;
    }

    @Override
    public ConnectorSetupPanel getEditSetupPanel(Sandbox sandbox, FileSetup setup) {
        SavableComponent savableComponent = sandbox.allowLocalFSAccess() ?
                new LocalModeFilePanel(setup) : createServerModePanel(sandbox, setup);

        return new ConnectorSetupPanel() {
            @Override
            public Component getComponent() {
                return savableComponent.getComponent();
            }

            @Override
            public Optional<String> validate() {
                return Optional.empty();
            }

            @Override
            public void showError(String string) {

            }

            @Override
            public ConnectorSetup getResult() {
                savableComponent.save();
                return setup;
            }
        };
    }

    @Override
    public FileSetup createDefaultSetup(Sandbox sandbox) {
        var newPath = FileNameGenerator.findSafeAvailableFileName(sandbox.getUserContentDirectory(), "MSP_%d.xml").getAbsolutePath();
        var label = getShortLabel(newPath);
        return FileSetup.apply(MSPConnector.ID, label, newPath, newPath);
    }

    @Override
    public List<BadConfigException> validateForSave(MSPConfig config, FileSetup setup, List<FieldMapping<?>> fieldMappings) {
        // empty target file name is valid because it will be generated in [[updateForSave]]
        // right before the export
        return Collections.emptyList();
    }

    @Override
    public List<BadConfigException> validateForLoad(MSPConfig config, FileSetup setup) {
        var seq = new ArrayList<BadConfigException>();
        if (setup.getSourceFile().isEmpty()) {
            seq.add(new InputFileNameNotSetException());
        }
        return seq;
    }

    @Override
    public void validateForDropInLoad(MSPConfig config) throws BadConfigException, DroppingNotSupportedException {
        // Always valid!
    }

    @Override
    public String describeSourceLocation(MSPConfig config, FileSetup setup) {
        return new File(setup.getSourceFile()).getName();
    }

    @Override
    public String describeDestinationLocation(MSPConfig config, FileSetup setup) {
        return new File(setup.getTargetFile()).getName();
    }

    @Override
    public Messages fieldNames() {
        return messages;
    }

    private static String getShortLabel(String fileName) {
        return Paths.get(fileName).getFileName().toString();
    }

    private SavableComponent createServerModePanel(Sandbox sandbox, FileSetup fileSetup) {
        return new ServerModeFilePanel(sandbox.getUserContentDirectory(), fileSetup,
                uploadedFile -> processFile(sandbox, uploadedFile)
        );
    }


    private FileProcessingResult processFile(Sandbox sandbox, File uploadedFile) {
        var fileName = uploadedFile.getName();
        // check if MPP file
        var isMpp = fileName.toLowerCase().endsWith(MSPFileReader.MPP_SUFFIX_LOWERCASE);
        if (!isMpp) {
            return new FileProcessingResult(uploadedFile, UPLOAD_SUCCESS);
        }
        var f = new File(sandbox.getUserContentDirectory(), fileName);
        var newFilePath = MSPUtils.convertMppProjectFileToXml(f.getAbsolutePath());
        if (newFilePath == null) return new FileProcessingResult(null, SAVE_FILE_FAILED);
        return new FileProcessingResult(new File(newFilePath), UPLOAD_MPP_SUCCESS);
    }
}

File: connectors-editors/msp-editor/src/main/scala/com/taskadapter/connector/msp/editor/MSPEditorFactory.scala
package com.taskadapter.connector.msp.editor

import com.taskadapter.common.ui.FieldMapping
import com.taskadapter.connector.common.FileNameGenerator
import com.taskadapter.connector.definition.FileSetup
import com.taskadapter.connector.definition.exceptions.BadConfigException
import com.taskadapter.connector.msp._
import com.taskadapter.connector.msp.editor.error.{InputFileNameNotSetException, OutputFileNameNotSetException}
import com.taskadapter.web.configeditor.file.{FileProcessingResult, LocalModeFilePanel, ServerModeFilePanel}
import com.taskadapter.web.data.Messages
import com.taskadapter.web.service.Sandbox
import com.taskadapter.web.uiapi.{DefaultSavableComponent, SavableComponent}
import com.taskadapter.web.{ConnectorSetupPanel, PluginEditorFactory}
import com.vaadin.flow.component.Component
import com.vaadin.flow.component.orderedlayout.VerticalLayout

import java.io.File
import java.nio.file.Paths
import java.util.{Collections, Optional}
import scala.collection.JavaConverters._
import scala.collection.mutable

class MSPEditorFactory extends PluginEditorFactory[MSPConfig, FileSetup] {
  private val BUNDLE_NAME = "com.taskadapter.connector.msp.messages"
  private val LABEL_DESCRIPTION_TEXT = "Description:"
  private val LABEL_TOOLTIP = "Text to show for this connector on 'Export' button. Enter any text."
  private val UPLOAD_SUCCESS = "Upload success"
  private val UPLOAD_MPP_SUCCESS = "File uploaded and successfully converted to XML"
  private val SAVE_FILE_FAILED = "Save file error" // error of saving after upload

  private val messages = new Messages(BUNDLE_NAME)

  override def formatError(e: Throwable): String = {
    e match {
      case relationType: UnsupportedRelationType => return messages.format("errors.unsupportedRelation",
        messages.get("relations." + relationType.getRelationType))
      case _: InputFileNameNotSetException => return messages.get("error.inputFileNameNotSet")
      case _: OutputFileNameNotSetException => return messages.get("error.outputFileNameNotSet")
      case _ => e.getMessage
    }
  }

  // TODO 14 use the binder
  override def getMiniPanelContents(sandbox: Sandbox, config: MSPConfig, setup: FileSetup): SavableComponent = {
    val layout = new VerticalLayout
    layout.setMargin(true)
    layout.add(createInfoReadOnlyPanel)
    // TODO 14 update the save function
    new DefaultSavableComponent(layout, () => {
      true
    })
  }

  override def getEditSetupPanel(sandbox: Sandbox, setup: FileSetup) = new ConnectorSetupPanel() {
    var savableComponent : SavableComponent = if (sandbox.allowLocalFSAccess)
      new LocalModeFilePanel(setup)
    else
      createServerModePanel(sandbox, setup)

    override def getComponent: Component = {
      savableComponent.getComponent
    }

    override def validate(): Optional[String] = {
      Optional.empty()
    }

    override def getResult: FileSetup = {
      savableComponent.save()
      setup
    }

    override def showError(String: String): Unit = {
      // TODO show error
    }
  }

  def getShortLabel(fileName: String): String = {
    Paths.get(fileName).getFileName.toString
  }

  private def createInfoReadOnlyPanel = {
    val infoPanel = new MSPInfoPanel
    infoPanel.setHeight("152px")
    infoPanel
  }

  private def createServerModePanel(sandbox: Sandbox, fileSetup: FileSetup): SavableComponent = {
    new ServerModeFilePanel(sandbox.getUserContentDirectory, fileSetup,
      (uploadedFile: File) => processFile(sandbox, uploadedFile)
    )
  }

  override def validateForSave(config: MSPConfig, setup: FileSetup, fieldMappings: java.util.List[FieldMapping[_]]):
  java.util.List[BadConfigException] = {
    // empty target file name is valid because it will be generated in [[updateForSave]]
    // right before the export
    Collections.emptyList();
  }

  override def validateForLoad(config: MSPConfig, setup: FileSetup): java.util.List[BadConfigException] = {
    val seq = new mutable.ListBuffer[BadConfigException]()
    if (setup.getSourceFile.isEmpty) seq += new InputFileNameNotSetException
    seq.asJava
  }

  override def validateForDropInLoad(config: MSPConfig): Unit = {
    // Always valid!
  }

  override def describeSourceLocation(config: MSPConfig, setup: FileSetup): String =
    new File(setup.getSourceFile).getName

  override def describeDestinationLocation(config: MSPConfig, setup: FileSetup): String =
    new File(setup.getTargetFile).getName

  private def processFile(sandbox: Sandbox, uploadedFile: File): FileProcessingResult = {
    val fileName = uploadedFile.getName
    // check if MPP file
    val isMpp = fileName.toLowerCase.endsWith(MSPFileReader.MPP_SUFFIX_LOWERCASE)
    if (!isMpp) return new FileProcessingResult(uploadedFile, UPLOAD_SUCCESS)
    val f = new File(sandbox.getUserContentDirectory, fileName)
    val newFilePath = MSPUtils.convertMppProjectFileToXml(f.getAbsolutePath)
    if (newFilePath == null) return new FileProcessingResult(null, SAVE_FILE_FAILED)
    new FileProcessingResult(new File(newFilePath), UPLOAD_MPP_SUCCESS)
  }

  override def isWebConnector: Boolean = false

  override def createDefaultSetup(sandbox: Sandbox): FileSetup = {
    val newPath = FileNameGenerator.findSafeAvailableFileName(sandbox.getUserContentDirectory, "MSP_%d.xml").getAbsolutePath
    if (newPath == null) throw new OutputFileNameNotSetException
    val label = getShortLabel(newPath)
    FileSetup.apply(MSPConnector.ID, label, newPath, newPath)
  }

  override def fieldNames: Messages = messages
}
